src/tactic/polyrith.lean

```diff
@@ -0,0 +1,605 @@
+* `(hyps : list pexpr)` - the hypotheses/proof terms selecteed by the user
+
+First, the tactic converts the target into a `poly`, and finds out what type it
+is an equality of. (It also fills up an `exmap` with its information). Then, it
+collects all the relevant hypotheses/proof terms from the context, and from those
+selected by the user, taking into account whether `only_on` is true. (The `exmap` is
+updated accordingly as well).
+
+This information is used to create a list of args that get used in a call to
+the appropriate python file that executes a grobner basis computation. The
+output of this computation is a `string` representing the certificate. This
+string is parsed into a list of `poly` objects that are then converted into
+`pexpr`s (using the updated `exmap`).
+
+the names of the hypotheses, along with the corresponding coefficients are
+given to `linear_combination`. If that tactic succeeds, the user is prompted
+to replace the call to `polyrith` with the appropriate call to
+`linear_combination`.
+-/
+meta def tactic.polyrith (only_on : bool) (hyps : list pexpr): tactic unit :=
+do
+  sleep 10, -- otherwise can lead to weird errors when actively editing code with polyrith calls
+  (m, p, R) ← parse_target_to_poly,
+  (eq_names, m, polys) ← parse_ctx_to_polys R m only_on hyps,
+  let args := [to_string R, (get_var_names m).to_string,
+    (polys.map poly.mk_string).to_string, p.mk_string],
+  let args := to_string (is_trace_enabled_for `polyrith) :: args,
+  sage_out ← sage_output args,
+  if is_trace_enabled_for `polyrith then
+    trace sage_out
+  else do
+  coeffs_as_poly ← convert_sage_output sage_out,
+  coeffs_as_pexpr ← coeffs_as_poly.mmap (poly.to_pexpr m),
+  let eq_names_pexpr := eq_names.map to_pexpr,
+  coeffs_as_expr ← coeffs_as_pexpr.mmap $ λ e, to_expr ``(%%e : %%R),
+  linear_combo.linear_combination eq_names_pexpr coeffs_as_pexpr,
+  let components := (eq_names.zip coeffs_as_expr).filter
+    $ λ pr, bnot $ pr.2.is_app_of `has_zero.zero,
+  expr_string ← components_to_lc_format components,
+  let cmd : format := "linear_combination " ++ format.nest 2 (format.group expr_string),
+  trace!"Try this: {cmd}"
+
+
+/-! # Interactivity -/
+setup_tactic_parser
+
+/--
+Attempts to prove polynomial equality goals through polynomial arithmetic
+on the hypotheses (and additional proof terms if the user specifies them).
+It proves the goal by generating an appropriate call to the tactic
+`linear_combination`. If this call succeeds, the call to `linear_combination`
+is suggested to the user.
+
+* `polyrith` will use all relevant hypotheses in the local context.
+* `polyrith [t1, t2, t3]` will add proof terms t1, t2, t3 to the local context.
+* `polyrith only [h1, h2, h3, t1, t2, t3]` will use only local hypotheses
+  `h1`, `h2`, `h3`, and proofs `t1`, `t2`, `t3`. It will ignore the rest of the local context.
+
+Note: This tactic only works with a working internet connection.
```

Comment:
````suggestion
Note: This tactic only works with a working internet connection.

Examples:

```
example (x y : ℚ) (h1 : x*y + 2*x = 1) (h2 : x = y) :
  x*y = -2*y + 1 :=
by polyrith
-- Try this: linear_combination h1 - 2 * h2

example (x y z w : ℚ) (hzw : z = w) : x*z + 2*y*z = x*w + 2*y*w :=
by polyrith
-- Try this: linear_combination (2 * y + x) * hzw

constant scary : ∀ a b : ℚ, a + b = 0

example (a b c d : ℚ) (h : a + b = 0) (h2: b + c = 0) : a + b + c + d = 0 :=
by polyrith only [scary c d, h]
-- Try this: linear_combination scary c d + h
```
````